File: src/main/java/com/temelyan/pomoapp/model/AbstractEntity.java
package com.temelyan.pomoapp.model;

import com.temelyan.pomoapp.HasId;

import javax.persistence.*;
import java.util.Objects;

@MappedSuperclass
@Access(AccessType.FIELD)
public abstract class AbstractEntity implements HasId {
    private static final int START_SEQ = 100000;

    @Id
    @SequenceGenerator(name = "global_seq", sequenceName = "global_seq", allocationSize = 1, initialValue = START_SEQ)
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "global_seq")
    @Access(value = AccessType.PROPERTY)
    private Integer id;

    protected AbstractEntity() {
    }

    AbstractEntity(Integer id) {
        this.id = id;
    }

    public Integer getId() {
        return id;
    }

    public void setId(Integer id) {
        this.id = id;
    }

    @Override
    public String toString() {
        return String.format("%s id = %s", getClass().getSimpleName(), getId());
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof AbstractEntity)) return false;
        AbstractEntity that = (AbstractEntity) o;
        return Objects.equals(id, that.id);
    }

    @Override
    public int hashCode() {
        return (getId() == null) ? 0 : getId();
    }
}

File: src/main/java/com/temelyan/pomoapp/model/AbstractEntity.kt
package com.temelyan.pomoapp.model

import com.temelyan.pomoapp.HasId
import javax.persistence.*

@MappedSuperclass
@Access(AccessType.FIELD)
abstract class AbstractEntity : HasId {

    @Id
    @SequenceGenerator(name = "global_seq", sequenceName = "global_seq", allocationSize = 1, initialValue = START_SEQ)
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "global_seq")
    @Access(value = AccessType.PROPERTY)
    private var id: Int? = null

    protected constructor()

    internal constructor(id: Int?) {
        this.id = id
    }

    override fun getId(): Int? {
        return this.id
    }

    final override fun setId(id: Int) {
        this.id = id
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is AbstractEntity) return false
        val that = other as AbstractEntity?
        return id == that!!.id
    }

    override fun hashCode(): Int {
        return getId() ?: 0
    }

    override fun toString(): String {
        return String.format("%s id = %s", javaClass.simpleName, getId())
    }

    companion object {
        private const val START_SEQ = 100000
    }
}
